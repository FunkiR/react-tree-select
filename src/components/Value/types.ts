import {ComponentType} from 'react';
import type {Value} from '~/types';

export type Props = {
	onClear: () => void,
	onToggleMenu: () => void,
	value: Value | null
};

export type ValueComponent = ComponentType<Props>;
