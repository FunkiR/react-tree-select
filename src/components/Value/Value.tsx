import {faCaretDown, faTimesCircle} from "@fortawesome/free-solid-svg-icons";
import useSelect from "~/hooks/useSelect";
import React, {memo} from "react";

import type {ValueComponent} from './types';

export const Value: ValueComponent = props => {
	const {onClear, onToggleMenu, value} = props;
	const [components, getters, theme] = useSelect('components', 'getters', 'theme');
	const {CaretIcon, ClearIcon, IndicatorsContainer, ValueContainer, ValueLabel} = components;
	const {label} = getters.value(value);

	return (
		<ValueContainer className={theme.ValueContainer}>
			<ValueLabel className={theme.ValueLabel} title={label}>
				{label}
			</ValueLabel>
			<IndicatorsContainer className={theme.IndicatorsContainer}>
				{value && <ClearIcon className={theme.ClearIcon} icon={faTimesCircle} onClick={onClear} />}
				<CaretIcon className={theme.CaretIcon} icon={faCaretDown} onClick={onToggleMenu} />
			</IndicatorsContainer>
		</ValueContainer>
	);
};

export default memo(Value);
