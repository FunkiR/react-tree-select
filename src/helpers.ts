import cn from 'classnames';

import withOriginalComponent from "./HOCs/withOriginalComponent";
import {AnyComponents, CustomTheme, SpreadObject, Theme, ThemeMode, ThemeOptions, ThemeWithOptions} from "./types";

/**
 * HOC to prevent frequent calls to passed function
 *
 * @template T
 *
 * @param {T} f - target function
 * @param {number} ms - time limit for repeat call check
 * @returns {T} - decorator with re-invocation check
 */
const debounce = <T extends Function>(f: T, ms: number): T => {
	let timer: NodeJS.Timeout;

	const callable = (...args) => {
		timer && clearTimeout(timer);

		timer = setTimeout(() => f(...args), ms);
	};

	return <T>(<any>callable);
};

/**
 * Used to check if a string contains a passed substring
 *
 * @param {string} string - string that will be searched for
 * @param {string} substring - substring to search
 * @returns {boolean} - result of check
 */
const hasSubstring = (string: string, substring?: string | number): boolean => {
	let searchValue = substring;

	if (Number.isInteger(searchValue)) searchValue = searchValue?.toString();

	return Boolean(searchValue && string.search(searchValue.toString()) != -1);
};

/**
 * Used as mock or default prop
 *
 * @returns {void} - empty result
 */
const noop = (): any => undefined;

/**
 * Extends theme by replacing or adding new selectors. Also changes theme settings if they were passed
 *
 * @template T, NT, RT
 *
 * @param {T | CustomTheme<T> | ThemeWithOptions<T>} data - data of base theme
 * @param {NT | CustomTheme<NT> |ThemeWithOptions<NT> | void} newData - data to extend base theme
 * @returns {ThemeWithOptions<RT>} - data of new theme with applied changes
 */
const extendTheme = <T extends Theme>(
	data: T | ThemeWithOptions<T>,
	newData?: CustomTheme<T> | ThemeWithOptions<CustomTheme<T> | null>
): ThemeWithOptions<T> => {
	const getThemeWithOptions = <T>(data?: T | CustomTheme<T> | ThemeWithOptions<T | null>, defaultOptions?: Partial<ThemeOptions>): ThemeWithOptions<T> => {
		let options: Partial<ThemeOptions> = {
			mode: ThemeMode.REPLACE
		};
		let theme;

		if (Array.isArray(data)) {
			[theme, options] = data;
		} else if (data) theme = data;

		return [theme, {...defaultOptions, ...options}];
	};

	const [defaultTheme, defaultOptions] = getThemeWithOptions<T>(data);
	const [customTheme, customOptions] = getThemeWithOptions<CustomTheme<T>>(newData, defaultOptions);
	const {mode: themeMode = ThemeMode.REPLACE, prefix} = customOptions;
	let theme = defaultTheme;

	const getCNByMode = (className: string, newClassName: string, mode: ThemeMode): string => {
		return mode === ThemeMode.EXTEND ? cn(className, newClassName) : newClassName;
	};

	const getThemeByPrefix = <T>(theme: T, prefix: string): T => Object.keys(theme).reduce((newTheme, key) => {
		return newTheme[key] ? {...newTheme, [key]: `${prefix}${theme[key]}`} : newTheme;
	}, {} as T)

	const getThemeByCustom = <NT, CT>(theme: NT, customTheme: CustomTheme<CT>, mode: ThemeMode): typeof theme => {
		let newTheme = theme;

		Object.keys(customTheme).forEach(key => {
			const value = customTheme[key];
			const themeCN = theme[key];
			const className = typeof value === 'function' ? value(themeCN) : getCNByMode(themeCN, value, mode);

			newTheme = {...newTheme, [key]: className};
		});

		return newTheme;
	};

	if (prefix) {
		const prefixTheme = getThemeByPrefix(defaultTheme, prefix);

		theme = themeMode === ThemeMode.EXTEND
			? getThemeByCustom(defaultTheme, prefixTheme, themeMode)
			: prefixTheme;
	}

	if (customTheme) {
		theme = getThemeByCustom(theme, customTheme, themeMode);
	}

	return [theme, customOptions];
};

/**
 * Extends base object
 *
 * @template O, NO
 *
 * @param {O} obj - base object
 * @param {NO} newObj  - new object to extend base
 * @returns {SpreadObject<O, NO>} - merged object
 */
const extendObject = <O, NO>(obj: O, newObj: NO): SpreadObject<O, NO> => ({...obj, ...newObj});

/**
 * Extends a component object and applies a "withOriginalComponent" HOC to an already existing component
 *
 * @template C, NC
 *
 * @param {C} components - base object of components
 * @param {NC} newComponents - components for extend base object
 * @returns {SpreadObject<C, NC>} - new object of components with extended components
 */
const extendComponents = <C extends AnyComponents, NC extends AnyComponents>(components: C, newComponents?: NC): SpreadObject<C, NC> => {
	let extendedComponents = newComponents ?? {} as NC;

	newComponents && Object.keys(newComponents).forEach(name => {
		const originalComponent = components[name];
		const component = newComponents[name];

		extendedComponents = {
			...extendedComponents,
			[name]: originalComponent ? withOriginalComponent(component, originalComponent) : component
		};
	});

	return extendObject(components, extendedComponents);
};

export {
	debounce,
	extendComponents,
	extendObject,
	extendTheme,
	hasSubstring,
	noop
};
