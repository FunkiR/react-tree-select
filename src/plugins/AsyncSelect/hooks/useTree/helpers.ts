import {AsyncNode, Getters, TreeMap} from "~/plugins/AsyncSelect/types";
import {Node as ArrayNode} from '~/types';

const getTreeMap = (tree: TreeMap | ArrayNode[], parent: string | null = null, getters: Getters)
	: TreeMap => Array.isArray(tree) ? arrayToTree(tree, parent, getters) : tree;

const getRoots = (map: TreeMap, getters: Getters): Array<string> =>
{
	return Object.keys(map).filter(value => !getters.node(map[value]).parent);
};

const arrayToTree = (arrayTree: Array<ArrayNode>, parentId: string | null = null, getters: Getters): TreeMap => {
	const tree = {};

	const getChildValue = (parent: string, nodeValue: string) => `${parent}$${nodeValue}`;

	const getTreeNodes = (arrayNode: ArrayNode, parent: string | null = null): Array<AsyncNode> => {
		const {children, hasChildren, label, value} = getters.arrayNode(arrayNode);
		const id = parent ? getChildValue(parent, value) : value;
		const node = {
			children: children.map(child => getChildValue(id, getters.arrayNode(child).value)),
			data: arrayNode,
			error: false,
			id,
			label,
			loading: false,
			hasChildren: (Array.isArray(children) && children.length > 0) || hasChildren,
			parent,
			uploaded: children.length > 0,
			value
		};

		return children
			.map(child => getTreeNodes(child, node.value))
			.reduce((nodes, childNodes) => [...nodes, ...childNodes], [node]);
	};

	arrayTree.forEach(node => {
		getTreeNodes(node, parentId).forEach(asyncNode => {
			tree[getters.node(asyncNode).id] = asyncNode;
		});
	});

	return tree;
};

export {
	arrayToTree,
	getRoots,
	getTreeMap
};
