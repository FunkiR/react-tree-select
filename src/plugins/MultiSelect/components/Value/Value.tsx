import {faCaretDown, faTimesCircle} from "@fortawesome/free-solid-svg-icons";
import useMultiSelect from "~/plugins/MultiSelect/hooks/useMultiSelect";
import useMultiSelectPlugin from "~/plugins/MultiSelect/hooks/useMultiSelectPlugin";
import React, {createRef, memo, useEffect, useState} from 'react';

import {MIN_TAG_WIDTH} from "./constants";
import type {ValueComponent} from "./types";

export const Value: ValueComponent = props => {
	const {onToggleMenu} = props;
	const [tagWidth, setTegWidth] = useState(0);
	const containerRef = createRef<HTMLDivElement>();
	const [handleClear, handleRemove, values] = useMultiSelectPlugin('handleClear', 'handleRemove', 'values');
	const [components, theme] = useMultiSelect('components', 'theme');
	const {CaretIcon, ClearIcon, IndicatorsContainer, Tags, ValueContainer} = components;

	useEffect(() => {
		const {current} = containerRef;

		if (current) {
			const {clientWidth: width} = current;
			const tagsCount = Math.floor(width / MIN_TAG_WIDTH);

			if (tagsCount > 0) {
				const newTagWidth = Math.floor(width / tagsCount);

				setTegWidth(Math.max(newTagWidth, MIN_TAG_WIDTH));
			}
		}
	}, []);

	const renderClearIcon = () => values.length > 0
		? <ClearIcon className={theme.ClearIcon} icon={faTimesCircle} onClick={handleClear} />
		: null;

	return (
		<ValueContainer className={theme.ValueContainer} ref={containerRef}>
			<Tags onRemove={handleRemove} tagWidth={tagWidth} values={values} />
			<IndicatorsContainer className={theme.IndicatorsContainer}>
				{renderClearIcon()}
				<CaretIcon className={theme.CaretIcon} icon={faCaretDown} onClick={onToggleMenu} />
			</IndicatorsContainer>
		</ValueContainer>
	);
};

export default memo(Value);
