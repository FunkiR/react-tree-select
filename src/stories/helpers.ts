import type {Node} from '~/types';

const generateArrayTree = (limit = 5,  levelsCount = 2, offset = 0): Node[] => {
	const createNode = (index, levelsCount) => ({
		label: `node${index + offset}`,
		value: `node${index + offset}`,
		children: levelsCount > 0 ? createNodes(levelsCount) : [],
	});

	const createNodes = (levelsCount = 1) => {
		const nextLevelsCount = levelsCount - 1;

		return Array.from(Array(Math.round(limit / 2), limit)).map((v, i) => {
			return createNode(i, nextLevelsCount);
		});
	};

	return createNodes(levelsCount);
};

export {
	generateArrayTree
};
