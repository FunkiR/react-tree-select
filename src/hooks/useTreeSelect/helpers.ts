import {extendComponents, extendObject, extendTheme} from "helpers";
import {AnyProps} from "~/types";

import {ApplyPluginReturnType} from './types';

const applyPlugin = <P extends AnyProps, PP extends AnyProps>(props: P, pluginProps: PP): ApplyPluginReturnType<P, PP> => {
	const {components, getters, setters, theme} = props;
	const {components: pluginComponents, getters: pluginGetters, setters: pluginSetters, theme: pluginTheme} = pluginProps;

	return {
		...props,
		...pluginProps,
		components: extendComponents(components, pluginComponents),
		getters: extendObject(getters, pluginGetters),
		setters: extendObject(setters, pluginSetters),
		theme: extendTheme(theme, pluginTheme)
	};
};

export {
	applyPlugin
};