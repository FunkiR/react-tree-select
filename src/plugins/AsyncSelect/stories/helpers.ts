import {InternalOnFetchEvent, InternalOnFetchReturnType} from "~/plugins/AsyncSelect/components/Select/types";
import type {DefaultArrayNode} from "~/plugins/AsyncSelect/types";
import {generateArrayTree} from "stories/helpers";

/**
 * Pauses code execution
 *
 * @param {number} ms - suspension time
 * @returns {Promise<void>}
 */
const sleep = (ms = 0): Promise<void> => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Returns random boolean value
 *
 * @returns {boolean} - random value
 */
const randomBool = (): boolean => Math.random() > 0.5;

/**
 * Generates an array tree with "hasChildren" property and offset relative to previously obtained nodes
 *
 * @param {number} offset - count of previously obtained nodes
 * @returns {DefaultArrayNode[]} - array tree
 */
const genAsyncArrayTree = (offset = 0): DefaultArrayNode[] => {
	const tree = generateArrayTree(10, 0, offset);

	return tree.map(node => ({...node, hasChildren: randomBool()}) as DefaultArrayNode);
};

/**
 * Returns an array tree given the offset
 *
 * @param {number} offset - count of previously obtained nodes
 * @returns {InternalOnFetchReturnType} - response for useTree hook
 */
const fetchData = async ({offset}: InternalOnFetchEvent): Promise<InternalOnFetchReturnType> => {
	await sleep(1000);
	const data = genAsyncArrayTree(offset);

	return {
		data,
		total: randomBool() ? Infinity : data.length
	};
};

export {
	fetchData,
	genAsyncArrayTree,
	sleep
};