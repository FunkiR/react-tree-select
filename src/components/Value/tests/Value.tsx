import {Value} from "components/Value/Value";
import {shallow} from "enzyme";
import {noop} from "helpers";
import React from "react";

describe('Value component', () => {
	const value = {label: 'label', value: 'value'};

	it('renders correctly', () => {
		const wrapper = shallow(<Value onClear={noop} onToggleMenu={noop} value={null} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders label equal to label of value', () => {
		const wrapper = shallow(<Value onClear={noop} onToggleMenu={noop} value={value} />);
		expect(wrapper.text()).toEqual(value.label);
	});

	it('handles onClear and call provided callback', () => {
		const onClear = jest.fn();
		const mockClearIcon = (): any => null;
		const useSelect = jest.requireMock('hooks/useSelect');

		useSelect.mockImplementationOnce((...keys) => {
			const values = useSelect(...keys);
			const index = keys.findIndex(key => key === 'components');

			if (index > -1) {
				values[index] = {...values[index], ClearIcon: mockClearIcon};
			}

			return values;
		});

		const wrapper = shallow(<Value onClear={onClear} onToggleMenu={noop} value={value} />);

		wrapper.find(mockClearIcon).prop('onClick')();
		expect(onClear.mock.calls.length).toEqual(1);
	});
});

