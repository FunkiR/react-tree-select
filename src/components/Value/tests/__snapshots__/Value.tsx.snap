// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value component renders correctly 1`] = `
<Memo(ForwardRef)
  className="ValueContainer"
>
  <Memo(ForwardRef)
    className="ValueLabel"
    title=""
  />
  <Memo(ForwardRef)
    className="IndicatorsContainer"
  >
    <Memo(Icon)
      className=""
      icon={
        Object {
          "icon": Array [],
          "iconName": "caret-down",
          "prefix": "fas",
        }
      }
      onClick={[Function]}
    />
  </Memo(ForwardRef)>
</Memo(ForwardRef)>
`;
